spring-boot-actuator/src/main/java/org/springframework/boot/actuate/endpoint/mvc/HeapdumpMvcEndpoint.java

```diff
@@ -0,0 +1,179 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.boot.actuate.endpoint.mvc;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.lang.management.ManagementFactory;
+import java.lang.reflect.Method;
+import java.text.SimpleDateFormat;
+import java.util.Collections;
+import java.util.Date;
+import java.util.Map;
+import java.util.zip.GZIPOutputStream;
+
+import javax.servlet.http.HttpServletResponse;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.util.ReflectionUtils;
+import org.springframework.util.StreamUtils;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+
+/**
+ * {@link MvcEndpoint} to expose heap dumps.
+ *
+ * @author Lari Hotari
+ */
+@ConfigurationProperties("endpoints.heapdump")
+@HypermediaDisabled
+public class HeapdumpMvcEndpoint extends AbstractMvcEndpoint implements MvcEndpoint {
+	private static final Log logger = LogFactory.getLog(HeapdumpMvcEndpoint.class);
+	private final Heapdumper heapdumper = new Heapdumper();
+	String fileNameBase = "heapdump";
+	private final ResponseEntity<Map<String, String>> HEAPDUMPER_NOT_AVAILABLE_RESPONSE = new ResponseEntity<Map<String, String>>(
+			Collections.singletonMap("message",
+					"Heapdumping is not supported in this environment."),
+			HttpStatus.NOT_FOUND);
+
+	public HeapdumpMvcEndpoint() {
+		setPath("/heapdump");
+	}
+
+	@RequestMapping(path = "", method = RequestMethod.GET)
+	public ResponseEntity<Map<String, String>> dump(HttpServletResponse response)
+			throws IOException {
+		return dumpHeap(response, false);
+	}
+
+	@RequestMapping(path = "/live", method = RequestMethod.GET)
+	public ResponseEntity<Map<String, String>> dumpLive(HttpServletResponse response)
+			throws IOException {
+		return dumpHeap(response, true);
+	}
+
+	private ResponseEntity<Map<String, String>> dumpHeap(HttpServletResponse response,
+			boolean live) throws IOException {
+		if (!isEnabled()) {
+			return MvcEndpoint.DISABLED_RESPONSE;
+		}
+		if (!heapdumper.isAvailable()) {
+			return HEAPDUMPER_NOT_AVAILABLE_RESPONSE;
+		}
+		final File dumpFile = File.createTempFile(createFileName(live), ".hprof");
+		// file must not exist before creating heap dump
+		dumpFile.delete();
+		try {
+			heapdumper.dumpHeap(dumpFile, live);
+			streamFileAndGzipToResponse(response, dumpFile);
+		}
+		finally {
+			dumpFile.delete();
+		}
+		return null;
+	}
+
+	private void streamFileAndGzipToResponse(HttpServletResponse response, File dumpFile)
+			throws IOException {
+		response.setContentType("application/octet-stream");
```

Comment:
The main reason for not doing that is to be as efficient as possible. Heapdumps compress a lot with Gzip and it's not possible to known the size of the gzipped content unless temporary files are used. When running an app in CloudFoundry there isn't a lot of diskspace available and that's why temporary disk space usage should be minimized as well. The streaming of the heap dump is in a try-finally block so that old files don't get left behind and consume the limited disk space.